streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -730,16 +827,150 @@ private static void buildAssignedActiveTaskAndPartitionsList(final List<TaskId>
+
+        // initialize task list for consumers
+        for (final String consumer : consumers) {
+            assignments.put(consumer, new ArrayList<>());
+        }
+
+        for (final TaskId task : state.activeTasks()) {
+            final Set<String> previousConsumers = previousConsumersOfTaskPartitions(partitionsForTask.get(task), state.ownedPartitions(), allOwnedPartitions);
+
+            // If this task's partitions were owned by different consumers, we can't avoid revoking partitions
+            if (previousConsumers.size() > 1) {
+                log.error("The partitions of task {} were claimed as owned by different StreamThreads. " +
+                    "This indicates the mapping from partitions to tasks has changed!", task);
+                return Collections.emptyMap();
+            }
+
+            // If this is a new task, or its old consumer no longer exists, it can be freely (re)assigned
+            if (previousConsumers.isEmpty()) {
+                log.debug("Task {} was not previously owned by any consumers still in the group. It's owner may " +
+                    "have died or it may be a new task", task);
+                newTasks.add(task);
+            } else {
+                final String consumer = previousConsumers.iterator().next();
+
+                // If the previous consumer was from another client, these partitions will have to be revoked
+                if (!consumers.contains(consumer)) {
+                    log.debug("This client was assigned a task {} whose partition(s) were previously owned by another " +
+                        "client, falling back on an interleaved assignment since a rebalance is inevitable.", task);
+                    return Collections.emptyMap();
+                }
+
+                // If this consumer previously owned more tasks than it has capacity for, some must be revoked
+                if (assignments.get(consumer).size() >= maxTasksPerClient) {
+                    return Collections.emptyMap();
+                }
+
+                assignments.get(consumer).add(task);
+
+                // If we have now reached capacity, remove it from set of consumers who still need more tasks
+                if (assignments.get(consumer).size() == maxTasksPerClient) {
+                    unfilledConsumers.remove(consumer);
+                }
+            }
+        }
+
+        // Interleave any remaining tasks among the consumers with remaining capacity
+        Collections.sort(newTasks);
+        while (!newTasks.isEmpty()) {
+            if (unfilledConsumers.isEmpty()) {
+                throw new IllegalStateException("Some tasks could not be distributed");
+            }
+
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            // Loop through the unfilled consumers and distribute tasks
+            while (consumerIt.hasNext()) {
```

Comment:
I think the logic here is not very intuitive as we are doing a double loop here. What we could do is to loop through all the unfilled consumers and fetch unfulfilled amount of new tasks to fed them. This could also reduce the number of loop cycles we have to go through.

Reply:
Well, the goal here is to interleave the tasks of the same groupId (ie subtopology) across different consumers as much as possible, since some subtopologies might be quite heavy while others are quite light. (This is what we do in `interleaveTasksByGroupId` also)
It is a little unintuitive I agree, so let me know if you have any suggestions for clearer code and/or comments. But if it's an consolation,
a) we probably aren't looping through that many remaining consumers and/or tasks
b) we may be doing a nested loop but within the inner loop we are removing things from the outer loop. So, in the end we are actually only looping over `newTasks` and hitting each task in it once

Reply:
I see your point, could you reflect your goal as meta comments for `interleaveTasksByGroupId`? It's a bit hard to interpret the goal just by reading this code.

Reply:
Sure, I'll try to shore up the explanation for `interleaveTasksByGroupId` and then refer to that here